helix-term/src/commands/lsp.rs

```diff
@@ -452,43 +452,73 @@ pub fn code_action(cx: &mut Context) {
     cx.callback(
         future,
         move |editor, compositor, response: Option<lsp::CodeActionResponse>| {
-            let actions = match response {
+            let mut actions = match response {
                 Some(a) => a,
                 None => return,
             };
+
             if actions.is_empty() {
                 editor.set_status("No code actions available");
                 return;
             }
 
-            let mut picker = ui::Menu::new(actions, (), move |editor, code_action, event| {
-                if event != PromptEvent::Validate {
-                    return;
+            // sort by CodeActionKind
+            // this ensures that the most relevant codeactions (quickfix) show up first
+            // while more situational commands (like refactors) show up later
+            // this behaviour is modeled after the behaviour of vscode (https://github.com/microsoft/vscode/blob/eaec601dd69aeb4abb63b9601a6f44308c8d8c6e/src/vs/editor/contrib/codeAction/browser/codeActionWidget.ts)
+
+            actions.sort_by_key(|action| match &action {
+                lsp::CodeActionOrCommand::CodeAction(lsp::CodeAction {
+                    kind: Some(kind), ..
+                }) => {
```

Comment:
CodeAction has a `is_preferred` field that I think should take some priority in this sorting: https://docs.rs/lsp-types/latest/lsp_types/struct.CodeAction.html#structfield.is_preferred

(For example, edit out the `mut` on [this line](https://github.com/helix-editor/helix/blob/e16c63276082f1cf79c57ff2ed2d301fe3333a0e/helix-term/src/keymap.rs#L186) and open up code actions on `command`: "Import `crate::commands`" and "Implement default members" are also quickfixes that will take priority over "consider changing this to be mutable" unless we take `is_preferred` into account.)

Reply:
Ah thanks for poiting that out, I took a look at how vscode handels this.
They always sort by category but within those category the codeactions are sorted by whether they fix any diagnostics and whether they are marked as preffered. I have implemented the same behaviour here.
The details are documented in the sourcecode (including permalinks to the vscode reference)